features/include/pcl/features/impl/good.hpp

```diff
@@ -0,0 +1,666 @@
+  double interval_x = largest_side / number_of_bins; 
+  double interval_y = largest_side / number_of_bins; 
+  for (size_t i = 0; i < XOY_projected_view->points.size(); i++)
+  {
+    pcl::PointXYZ p;
+    p.x = sign * XOY_projected_view->points.at(i).x + x;
+    p.y = sign * XOY_projected_view->points.at(i).y + y;        
+    if ((trunc(p.x / interval_x) < XOY_histogram.size()) and (trunc(p.y / interval_y) < XOY_histogram.at(0).size()) and 
+	  (trunc(p.x / interval_x) >= 0) and (trunc(p.y / interval_y) >= 0))
+    {
+	XOY_histogram.at(trunc(p.x / interval_x)).at(trunc(p.y / interval_y))++;
+    }
+  }
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void
+pcl::GOODEstimation<PointInT, NumberOfBins>::normalizingHistogram (std::vector<unsigned int> histogram, std::vector<float> &normalized_histogram)
+{
+  int sum_all_bins = 0;    
+  //compute sumation of all histogram's bins.
+  for(std::vector<unsigned int>::iterator it = histogram.begin(); it != histogram.end(); ++it)
+  {
+    sum_all_bins += *it;
+  }
+  if (sum_all_bins != 0)
+  {  
+    for(std::vector<unsigned int>::iterator it = histogram.begin(); it != histogram.end(); ++it)
+    {
+      normalized_histogram.push_back (*it / float (sum_all_bins));
+    }  
+  }
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void
+pcl::GOODEstimation<PointInT, NumberOfBins>::viewpointEntropy (std::vector<float> normalized_histogram, float &entropy)
+{
+  //http://stats.stackexchange.com/questions/66108/why-is-entropy-maximised-when-the-probability-distribution-is-uniform
+  entropy =0;  
+  for(std::vector<float>::iterator it = normalized_histogram.begin(); it != normalized_histogram.end(); ++it)
+  {
+    if (*it != 0)
+    {
+      float entropy_tmp = *it * log2 (*it);
+      entropy += entropy_tmp;
+    }
+  }
+  entropy = -entropy;  
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void
+pcl::GOODEstimation<PointInT, NumberOfBins>::findMaxViewPointEntropy (std::vector <float> view_point_entropy, int &index)
+{
+  index = 0;
+  std::vector<float>::iterator it;
+  it=std::max_element (view_point_entropy.begin(), view_point_entropy.end());
+  index = it - view_point_entropy.begin ();
+  //std::cout << "\nindex ="<< it - view_point_entropy.begin() <<"\t , content = " << *it <<"\n";  
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void
+pcl::GOODEstimation<PointInT, NumberOfBins>::averageHistograms (std::vector<float> histogram1, std::vector<float> historam2, std::vector<float> historam3, std::vector<float> &average)
+{
+  for (size_t i = 0; i < histogram1.size(); i++ )
```

Comment:
Good style will be to resize `average` in the beginning, because you know what the number of elements will be in the end. This will avoid useless re-allocations while pushing back new items. Then you just access by index.